File: usecase/stream_usecase.go
package usecase

import (
	"context"
	"fmt"
	"io"
	"mime"
	"path/filepath"

	"github.com/teledrive/teledrive/domain"
	tginfra "github.com/teledrive/teledrive/infrastructure/telegram"
)

// StreamUsecase handles file upload and download streaming through Telegram.
type StreamUsecase struct {
	fileRepo   domain.FileRepository
	userRepo   domain.UserRepository
	clientPool *tginfra.ClientPool
	uploader   *tginfra.Uploader
	downloader *tginfra.Downloader
}

// NewStreamUsecase creates a new StreamUsecase.
func NewStreamUsecase(
	fileRepo domain.FileRepository,
	userRepo domain.UserRepository,
	clientPool *tginfra.ClientPool,
	uploader *tginfra.Uploader,
	downloader *tginfra.Downloader,
) *StreamUsecase {
	return &StreamUsecase{
		fileRepo:   fileRepo,
		userRepo:   userRepo,
		clientPool: clientPool,
		uploader:   uploader,
		downloader: downloader,
	}
}

// Upload uploads a file to Telegram and creates a file record in the database.
func (uc *StreamUsecase) Upload(
	ctx context.Context,
	userID int,
	folderID *int,
	fileName string,
	fileSize int64,
	reader io.Reader,
) (*domain.File, error) {
	// Get the user to access their session.
	user, err := uc.userRepo.GetByID(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	// Get a Telegram client from the pool.
	client, err := uc.clientPool.GetClient(userID, user.SessionData)
	if err != nil {
		return nil, fmt.Errorf("failed to get telegram client: %w", err)
	}

	// Upload the file to Telegram.
	result, err := uc.uploader.UploadFile(ctx, client, reader, fileName, fileSize)
	if err != nil {
		return nil, fmt.Errorf("failed to upload file: %w", err)
	}

	// Detect MIME type from extension.
	mimeType := mime.TypeByExtension(filepath.Ext(fileName))
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}

	// Create file record in DB.
	file := &domain.File{
		UserID:            userID,
		FolderID:          folderID,
		Name:              fileName,
		Size:              fileSize,
		MimeType:          mimeType,
		TelegramMessageID: result.MessageID,
		TelegramChatID:    result.ChatID,
		TelegramFileID:    result.FileID,
	}

	created, err := uc.fileRepo.Create(file)
	if err != nil {
		return nil, fmt.Errorf("failed to create file record: %w", err)
	}

	return created, nil
}

// DownloadResult contains metadata for a file download.
type DownloadResult struct {
	FileName string
	MimeType string
	FileSize int64
}

// Download streams a file from Telegram to the provided writer.
func (uc *StreamUsecase) Download(
	ctx context.Context,
	userID int,
	fileID int,
	writer io.Writer,
) (*DownloadResult, error) {
	// Get the file record.
	file, err := uc.fileRepo.GetByID(fileID)
	if err != nil {
		return nil, fmt.Errorf("file not found: %w", err)
	}
	if file.UserID != userID {
		return nil, fmt.Errorf("file does not belong to user")
	}

	// Get the user to access their session.
	user, err := uc.userRepo.GetByID(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	// Get a Telegram client from the pool.
	client, err := uc.clientPool.GetClient(userID, user.SessionData)
	if err != nil {
		return nil, fmt.Errorf("failed to get telegram client: %w", err)
	}

	// Download the file from Telegram.
	err = uc.downloader.DownloadFile(ctx, client, file.TelegramMessageID, file.TelegramChatID, writer)
	if err != nil {
		return nil, fmt.Errorf("failed to download file: %w", err)
	}

	return &DownloadResult{
		FileName: file.Name,
		MimeType: file.MimeType,
		FileSize: file.Size,
	}, nil
}

File: src/usecase/stream.usecase.ts
import mime from 'mime-types';
import type { Readable, Writable } from 'stream';
import type { File, FileRepository } from '../domain/file.js';
import type { UserRepository } from '../domain/user.js';
import type { ClientPool } from '../infrastructure/telegram/client-pool.js';
import type { Uploader } from '../infrastructure/telegram/uploader.js';
import type { Downloader } from '../infrastructure/telegram/downloader.js';

export interface DownloadResult {
  fileName: string;
  mimeType: string;
  fileSize: number;
}

export class StreamUsecase {
  private fileRepo: FileRepository;
  private userRepo: UserRepository;
  private clientPool: ClientPool;
  private uploader: Uploader;
  private downloader: Downloader;

  constructor(
    fileRepo: FileRepository,
    userRepo: UserRepository,
    clientPool: ClientPool,
    uploader: Uploader,
    downloader: Downloader
  ) {
    this.fileRepo = fileRepo;
    this.userRepo = userRepo;
    this.clientPool = clientPool;
    this.uploader = uploader;
    this.downloader = downloader;
  }

  public async upload(
    userId: number,
    folderId: number | null,
    fileName: string,
    fileSize: number,
    reader: Readable | Buffer
  ): Promise<File> {
    const user = await this.userRepo.getById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    const client = await this.clientPool.getClient(userId, user.session_data);
    const result = await this.uploader.uploadFile(client, reader, fileName, fileSize);

    const detectedMime = mime.lookup(fileName);
    const mimeType = detectedMime || 'application/octet-stream';

    const created = await this.fileRepo.create({
      user_id: userId,
      folder_id: folderId,
      name: fileName,
      size: fileSize,
      mime_type: mimeType,
      telegram_message_id: result.messageId,
      telegram_chat_id: result.chatId,
      telegram_file_id: result.fileId,
    });

    return created;
  }

  public async download(
    userId: number,
    fileId: number,
    writer: Writable
  ): Promise<DownloadResult> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    const user = await this.userRepo.getById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    const client = await this.clientPool.getClient(userId, user.session_data);
    await this.downloader.downloadFile(client, file.telegram_message_id, writer);

    return {
      fileName: file.name,
      mimeType: file.mime_type,
      fileSize: file.size,
    };
  }

  public async getFileMetadata(userId: number, fileId: number): Promise<File> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }
    return file;
  }
}
